presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1774,35 +1768,39 @@ else if (criteria instanceof JoinOn) {
             return output;
         }
 
-        private void verifyJoinOnConditionReferencesRelatedFields(Relation rightRelation, Expression expression)
+        private void verifyJoinOnConditionReferencesRelatedFields(Scope rightScope, Expression expression, Relation rightRelation)
+        {
+            if (!isJoinOnConditionReferencesRelatedFields(expression, rightScope)) {
+                Optional<String> tableName = tryGetTableName(rightRelation);
+                String warningMessage = tableName.isPresent()
+                        ?
+                        createWarningMessage(
+                                expression,
+                                format(
+                                        "JOIN ON condition(s) do not reference the joined table '%s' and other tables in the same " +
+                                                "expression that can cause performance issues as it may lead to a cross join with filter",
+                                        tableName.get()))
+                        :
+                        createWarningMessage(
+                                expression,
+                                "JOIN ON condition(s) do not reference the joined relation and other relation in the same " +
+                                        "expression that can cause performance issues as it may lead to a cross join with filter");
+                warningCollector.add(new PrestoWarning(PERFORMANCE_WARNING, warningMessage));
+            }
+        }
+
+        private Optional<String> tryGetTableName(Relation relation)
         {
-            if (!(rightRelation instanceof AliasedRelation || rightRelation instanceof Table)) {
-                // TODO Add handling of TableSubquery, Lateral and Table for the case when columns in
-                //  JOIN ON condition specified without table name or alias (See #17382).
-                return;
+            if (relation instanceof Table) {
+                return Optional.of(((Table) relation).getName().toString());
             }
-            if (!isJoinOnConditionReferencesRelatedFields(expression, rightRelation)) {
-                String rightTableName;
-                if (rightRelation instanceof Table) {
-                    rightTableName = ((Table) rightRelation).getName().toString();
+            else if (relation instanceof AliasedRelation) {
+                AliasedRelation aliasedRelation = (AliasedRelation) relation;
+                if (aliasedRelation.getRelation() instanceof Table) {
```

Comment:
We can call tryGetTable here on the aliasedRelation.getRelation()?

Reply:
I do not see `tryGetTable()` method anywhere, only `tryGetTableName()`. And you are pointing to the body of `tryGetTableName()`.

Reply:
yeah tryGetTableName, it could have kept the code cleaner. Anyways @highker merged the PR.